### DISCUSSION TITLE
`PathIntegrator` with different spps for different pixels

### DISCUSSION THREAD

**Discussion by treder** [mitsuba-renderer/mitsuba3#1178] [Q&A] [answered]
I'm quite new to Mitsuba and was looking for a way to have different `spp` for different pixels. For instance, I'd like to pass a [w x h] int array specifying spp to the `render` function. Any pointers for a Python implementation?

On a related note, is there a Python implementation of the `PathIntegrator`? It might be a useful starting point for custom path integrators.

**Comment by ziyi-zhang** [comment, marked as answer]
Hi,

This can be achieved by overriding the `sample_rays()` function in `common.py`.

For python implementation of path, `prb.py` is almost "path", with some extra complications of AD propagation that can be removed if not needed. 
Best

**Reply by treder** [reply]
Great, thank you for the pointers!
